### DISCUSSION TITLE
Can finalize hook be triggered when pod is deleted ?

### DISCUSSION THREAD

**Discussion by K-Jean** [metacontroller/metacontroller#368] [Q&A]
I have a service that should be accessible when a pod is in the "Terminating" state.  
For this I use the "publishNotReadyAddresses" on my service.
My problem is that when the pod goes into the "Terminating" state the finalize hook is called and my service is destroyed.

**Comment by grzesuav** [comment]
Hi,

not sure if I understand your use-case - how your operator works ? What does not work as expected ?

**Comment by grzesuav** [comment]
Moving to discussions for now

**Comment by grzesuav** [comment]
@K-Jean ok, I think I got what you mean.

When pod get's deleted, it gets a `deletionTimestamp` - which is indicator that pod is being deleted - however then pod's enter the termination state (depended on `terminationGracePeriodSeconds` etc) . This is indicator for kubernetes GarbageCollector to trigger resource removals by call all objects having finalizers. So it is expected behavior.

How exactly your pod is destroyed ? As metacontroller cannot remove/destroy the pod, it is the job of the kubelet.

**Reply by K-Jean** [reply]
Oh I forgot this ticket for so long ! Thanks for the re-up !

I have a pod which I need to keep active when is going to the `Terminating` state. For that I add a Service with the `publishNotReadyAddresses` flag. With this I can still use my pods when is under the `Terminating` state.

The problems is when I scaledown my service, the metacontroller hooks I triggered and my service is deleted before my pods :( 
What I need is that metacontroller wait the deletion of my pods before deleting the service associate with it. 

I will try to join an exemple of what I did with metacontroller.

**Reply by grzesuav** [reply]
@K-Jean do you use `sync` hook  only or also `finalize` hook ? Do you use `Decoration` or `Composite` controller ? What are children types ?

As I think, with finalize hook :
* it will be called when you delete parent resource
* you can (on first invocation) remove the pods from children and return just service
* return `false` so you will be re-invoked
* repeat until in request, in observed children you will notice no pods
